### PR TITLE
[ios] Fix the layout constraints on iOS 11 for iPhone X 

### PR DESCRIPTION
**Pull request by fabian-guerra**
Fixes https://github.com/mapbox/mapbox-gl-native/issues/10834

### CODE REVIEW

File: platform/ios/src/MGLMapView.mm [hunk 1]
```diff
@@ -914,17 +914,21 @@ - (void)updateConstraints
         // logo view
         [self removeConstraints:self.logoViewConstraints];
         [self.logoViewConstraints removeAllObjects];
-        [self.logoViewConstraints addObject:[safeAreaLayoutGuide.bottomAnchor constraintEqualToAnchor:self.logoView.bottomAnchor
-                                                                                             constant:8.0 + self.contentInset.bottom]];
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wpartial-availability"
+        [NSLayoutConstraint activateConstraints:@[[safeAreaLayoutGuide.bottomAnchor constraintEqualToSystemSpacingBelowAnchor:self.logoView.bottomAnchor multiplier:1.0]]];
```
Comment (friedbunny) [inline platform/ios/src/MGLMapView.mm]: This seems to accomplish the same thing as with `compassViewConstraints` above, but takes a different approach: activating the new constraint immediately. Is one way better/different than the other?

**Comment by friedbunny** [comment]
I agree with https://github.com/mapbox/mapbox-gl-native/issues/10834#issuecomment-355422584 in that we really need to add tests, otherwise we can’t consider this issue fixed.

File: platform/ios/src/MGLMapView.mm (hunk 1 above)
Comment (fabian-guerra) [reply platform/ios/src/MGLMapView.mm]: Looks like the correct way to add a bottom constraint on iOS 11 is anchor the view to the `safeAreaLayoutGuide` I was doing the opposite and it was not working.

```
[safeAreaLayoutGuide.bottomAnchor constraintEqualToSystemSpacingBelowAnchor:self.logoView.bottomAnchor multiplier:1.0]
```

**Comment by fabian-guerra** [comment]
@akitchen I bring your test case to this branch and did minor changes to pass the test prior iOS 11.

/cc @friedbunny

File: platform/ios/src/MGLMapView.mm (hunk 1 above)
Comment (akitchen) [reply platform/ios/src/MGLMapView.mm]: Sounds good. However, while we're in this code, can you refactor it to eliminate the duplication? Now that we have a test in place hopefully it feels less risky to do so.

File: platform/ios/test/MGLMapViewLayoutTests.m [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+- (void)tearDown {
+    self.styleLoadingExpectation = nil;
+    self.mapView = nil;
+
+    [super tearDown];
+}
+
+// TODO: check constraining to superview vs. constraining to safeAreaLayoutGuide.bottomAnchor
```
Comment (akitchen) [inline platform/ios/test/MGLMapViewLayoutTests.m]: I think you may have addressed this TODO? I thought we had an issue with layout depending on how the superview was constraining the map view. Please either delete this comment or test both cases (happy to pair on this if it's not clear)

File: platform/ios/test/MGLMapViewLayoutTests.m [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    CGFloat bottomSafeAreaInset = 0.0;
+    double accuracy = 0.01;
+
+    if ( [self.mapView respondsToSelector:@selector(safeAreaInsets)] ) {
+        bottomSafeAreaInset = self.mapView.safeAreaInsets.bottom;
+    }
+    
+    //compass
+    UIImageView *compassView = self.mapView.compassView;
+    NSLog(@"================> %@", compassView);
```
Comment (akitchen) [inline platform/ios/test/MGLMapViewLayoutTests.m]: Please remove all of the extraneous logging now that we don't need it, as it clutters the already noisy test output.

File: platform/ios/test/MGLMapViewLayoutTests.m [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    CGFloat expectedCompassOriginX = CGRectGetMaxX(self.mapView.bounds) - margin - CGRectGetWidth(compassView.frame);
+    CGFloat expectedCompassOriginY = margin;
+    // what about width/height? maybe we don't care as much at the moment?
```
Comment (akitchen) [inline platform/ios/test/MGLMapViewLayoutTests.m]: Let's either inspect width/height or remove the comment. This was just here as a placeholder while we were defining the test case.

**Comment by frederoni** [comment]
Have you considered adding a layer of abstraction on top of the layout API? It would remove a lot of the duplicated code, remove the need for separate code paths for `-[MGLMapView updateConstraintsPreiOS11]` and `-[MGLMapView updateConstraints]` and avoid the pushing and popping of diagnostic compiler flags.

I'm thinking something along these lines:
```objc
@implementation UIView (MGLAdditions)
- (NSLayoutYAxisAnchor *)safeTopAnchor {
    if (@available(iOS 11.0, *)) {
        return self.safeAreaLayoutGuide.topAnchor;
    } else {
        UIViewController *viewController = self.viewControllerForLayoutGuides;
        BOOL useLayoutGuides = viewController.view && viewController.automaticallyAdjustsScrollViewInsets;
        if (useLayoutGuides) {
            return viewController.topLayoutGuide.bottomAnchor;
        } else {
            return self.topAnchor;
        }
    }
}
@end
```
(One additional layer of abstraction is needed to support iOS 8)

**Comment by fabian-guerra** [comment]
@frederoni I didn't consider add another level of abstraction tbh. I refactored the code to remove the compiler flags pushing and popping. Since we are using two different APIs for managing the ornaments layout, seems that having two different methods is more clear to read aka `-[MGLMapView updateConstraintsPreiOS11]` and `-[MGLMapView updateConstraints]`

**Comment by akitchen** [comment]
> I refactored the code to remove the compiler flags pushing and popping.

This is great. Can you take it a bit further while we're here?

We require compiling against iOS 11 now, right? So I think we can also remove the code around https://github.com/mapbox/mapbox-gl-native/pull/10858/files#diff-20502b6bc02986a4398d0da84b2d4b94R884

 (cc @friedbunny)

**Comment by akitchen** [comment]
On second thought, since we want to cut version 3.7.3 let's get this merged in for the release and log a followup issue for code cleanup related to compile-time version checks.

File: platform/ios/src/MGLMapView.mm (hunk 1 above)
Comment (akitchen) [reply platform/ios/src/MGLMapView.mm]: > activating the new constraint immediately

I didn't quite understand - is this necessary @fabian-guerra ?

Comment (fabian-guerra) [reply platform/ios/src/MGLMapView.mm]: This code is no longer valid. Researching the reasons for why the constraint was not applied I tried to force activating the constraint that's why I added this code. If you take a second look you will see that I no longer use this.

**Pull request merged**